Add signing for mac osx.

Siedlerchr opened the pull request:
First part of fixing #6154 

- [x]  Sign and notarize dmg 
- [x]  Sign and notarize pkg

Edit: We maybe should run the signing only on master?


<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->


<!-- 
- Go through the list below. If a task has been completed, mark it done by using `[x]`.
- Please don't remove any items, just leave them unchecked if they are not applicable.
-->

- [ ] Change in CHANGELOG.md described (if applicable)
- [ ] Tests created for changes (if applicable)
- [ ] Manually tested changed features in running JabRef (always required)
- [ ] Screenshots added in PR description (for UI changes)
- [ ] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

Siedlerchr commented:
windows seems to fail:
Running [C:\Program Files (x86)\WiX Toolset v3.11\bin\light.exe, -nologo, -spdb, -ext, WixUtilExtension, -out, D:\a\jabref\jabref\build\distribution\JabRef-5.1.30845.msi, -sice:ICE27, -ext, WixUIExtension, -loc, D:\a\jabref\jabref\build\installer\config\MsiInstallerStrings_en.wxl, -b, D:\a\jabref\jabref\build\installer\config, D:\a\jabref\jabref\build\installer\wixobj\main.wixobj, D:\a\jabref\jabref\build\installer\wixobj\bundle.wixobj]in D:\a\jabref\jabref\build\installer\images\win-msi.image\JabRef
D:\a\jabref\jabref\build\installer\config\main.wxs(85) : error LGHT0091 : Duplicate symbol 'Property:ARPPRODUCTICON' found. This typically means that an Id is duplicated. Check to make sure all your identifiers of a given type (File, Component, Feature) are unique.
D:\a\jabref\jabref\build\installer\config\main.wxs(147) : error LGHT0092 : Location of symbol related to previous error.
java.io.IOException: Command [C:\Program Files (x86)\WiX Toolset v3.11\bin\light.exe, -nologo, -spdb, -ext, WixUtilExtension, -out, D:\a\jabref\jabref\build\distribution\JabRef-5.1.30845.msi, -sice:ICE27, -ext, WixUIExtension, -loc, D:\a\jabref\jabref\build\installer\config\MsiInstallerStrings_en.wxl, -b, D:\a\jabref\jabref\build\installer\config, D:\a\jabref\jabref\build\installer\wixobj\main.wixobj, D:\a\jabref\jabref\build\installer\wixobj\bundle.wixobj]in D:\a\jabref\jabref\build\installer\images\win-msi.image\JabRef exited with 92

Siedlerchr commented:
@Gena928  @k3KAW8Pnf7mkmdSMPHz27
Can you please test this  version here on mac?
The pgk and the dmg should now be signed correctly:
https://builds.jabref.org/pull/6748/merge/

(Just want to make sure that it works on your machines as well, because I have the certs already imported)

Gena928 commented:
Good day,
I was able to install JabRef using DMG & PKG files. 
Unfortunately icons in dmg file are not in the center in the install frame. 
<img width="532" alt="DMG install screen" src="https://user-images.githubusercontent.com/20572030/89740388-c7c5db80-da90-11ea-8fc4-53f2e49658ac.png">

But this looks much better now.

Siedlerchr commented:
@Gena928  Yeah! glad to hear that it works now fine.
Regarding the icons, I think we need to copy over that script and adapt the sizes:

```
  if theFilePath is "DEPLOY_INSTALL_LOCATION" then
      -- Position install location
      set position of item theFile of theWindow to {390, 130}
    else
      -- Position application or runtime
      set position of item theFile of theWindow to {120, 130}
    end if
```

https://github.com/openjdk/jdk/blob/master/src/jdk.incubator.jpackage/macosx/classes/jdk/incubator/jpackage/internal/resources/DMGsetup.scpt

Gena928 commented:
OK, let me know if you need my testing.
But I think the result is very good now. Nice work!

k3KAW8Pnf7mkmdSMPHz27 commented:
Awesome that you are working on this!

.pkg works as it should (signed and not notarized), but I don't know how to verify the signature of the .dmg.

`spctl -t open --assess --verbose JabRef-5.1.dmg` and `spctl -a -t open --context context:primary-signature -v JabRef-5.1.dmg` gives me
```
JabRef-5.1.dmg: rejected
source=Insufficient Context
```
and 
```
JabRef-5.1.dmg: rejected
source=Insufficient Context
```

but I am guessing this is as it should be right now?

Siedlerchr commented:
Thanks for the feedback, glad that the signing at least works. Notarization seems to be the next step after the build.

k3KAW8Pnf7mkmdSMPHz27 commented:
I get no error message related to signing from either the dmg or pkg.

The pkg displays the warning for it not being notarized ("... can't be opened because Apple cannot check it for malicious software.")

Siedlerchr commented:
Yeah, I noticed that too. I will try to setup the notarization workflow next. Then aftewards ist should be also possible to add it to the app store. Probably reqires another different certificate signing stuff

tobiasdiez commented on .github/workflows/deployment.yml:
```diff
@@ -55,14 +56,29 @@ jobs:
         uses: actions/setup-java@v1
         with:
-          java-version: 14
+          java-version: 15-ea
```
Why do we need to upgrade to 15?

Siedlerchr replied:
Because of this bug https://bugs.openjdk.java.net/browse/JDK-8238184

tobiasdiez commented on .github/workflows/deployment.yml:
```diff
@@ -55,14 +56,29 @@ jobs:
           key: ${{ runner.os }}-gradle-${{ hashFiles('**/*.gradle*') }}
           restore-keys: |
             ${{ runner.os }}-gradle-
+      - name: setup OSX Keychain on OSX only
```
Start with capitalized: `Setup` (here and below)

tobiasdiez commented on buildres/windows/JabRef-post-image.wsf:
```diff
@@ -23,9 +23,6 @@
         contents = contents.replace("</Product>", "<DirectoryRef Id=\"TARGETDIR\"><Component Id=\"RegistryJabRefBrowserEntries\" Guid=\"b6bc55ad-905c-4258-89b1-8b37abbe559c\" Win64=\"yes\"><RegistryKey Root=\"HKMU\" Key=\"SOFTWARE\\Mozilla\\NativeMessagingHosts\\org.jabref.jabref\" Action=\"createAndRemoveOnUninstall\" ForceCreateOnInstall=\"yes\"><RegistryValue Type=\"string\" Value=\"[INSTALLDI ...
 
-        // Specify correct icon in Add/Remove Programs
-        contents = contents.replace("</Product>", "<Property Id=\"ARPPRODUCTICON\" Value=\"DesktopIcon.exe\" /></Product>");
```
Why did you removed this?

Siedlerchr replied:
This workaround is no longer need, it has been fixed in 15 and the workaround prevented the windows build under 15
https://github.com/JabRef/jabref/pull/6748#issuecomment-671054116

tobiasdiez commented:
Concerning the installation screen, the easiest option is probably to change the image.

Siedlerchr commented:
I played a bit around with the a apple script, however, I still need to find out the value some of those "constants" , does not yet work. Best option would be to notarize the release manually.

tobiasdiez commented on buildres/mac/JabRef-dmg-setup.applescript:
```diff
@@ -0,0 +1,39 @@
+tell application "Finder"
```
Why do we need to include this script? Is it because of the image position? If yes, then please change the img instead. 

It's really ugly that one needs to copy-paste the installation scripts. In this way, we don't get the enhancments and fixes from the JDK. So if its possible, I would strongly advice not to copy paste them.

Siedlerchr replied:
It's actually not an installation script, it's just a macro script to set the background image and the icon position in the dmg. I figured out that I don't need to modify the variables. So we can always easily include upstream fixes.
It's explicity meant for overriding https://docs.oracle.com/en/java/javase/14/jpackage/override-jpackage-resources.html#GUID-62412365-821D-4736-A20D-70F2D47136D8

tobiasdiez replied:
So what did you actually changed then?

Siedlerchr replied:
<img width="686" alt="grafik" src="https://user-images.githubusercontent.com/320228/90321143-41554200-df47-11ea-93da-4c57373b0246.png">

I changed the window positions and the icon position, by moving the icon position a few pixes down and to the right.

tobiasdiez replied:
That's what I meant. You can also keep them in the standard location, and change the background image instead to adapt to the these locations.

Siedlerchr commented:
@Gena928  @k3KAW8Pnf7mkmdSMPHz27  I fixed the misaligned dmg images and also added JabRef volume icon and also the icon to the installer. Would be nice if you could show the screenshot of the dmg then.
Both are signed as before, but not yet notarized. This requires some extra steps in preparation, I will create a second PR as soon the prerequisites are

Gena928 commented:
Good day,
here you go my screen
<img width="926" alt="DMG install screen" src="https://user-images.githubusercontent.com/20572030/90321561-9dbe5e80-df53-11ea-8363-094e825cb237.png">

Siedlerchr commented:
Tested notarization, but does not yet work:

````
  "logFormatVersion": 1,
  "jobId": "7324997a-d63b-46e9-93b1-87330e7b0b73",
  "status": "Invalid",
  "statusSummary": "Archive contains critical validation errors",
  "statusCode": 4000,
  "archiveFilename": "JabRef-100.0.0.dmg",
  "uploadDate": "2020-08-16T17:17:19Z",
  "sha256": "4b78742a404109bbef7215daab00aca18a556877284ecf11d1e71fcaff030273",
  "ticketContents": null,
  "issues": [
    {
      "severity": "error",
      "code": null,
      "path": "JabRef-100.0.0.dmg/JabRef.app/Contents/MacOS/JabRef",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "x86_64"
    },
    {
      "severity": "error",
      "code": null,
      "path": "JabRef-100.0.0.dmg/JabRef.app/Contents/runtime/Contents/MacOS/libjli.dylib",
      "message": "The signature of the binary is invalid.",
      "docUrl": null,
      "architecture": "x86_64"
    }
  ]
````

Siedlerchr commented:
Bug Report: https://bugs.openjdk.java.net/browse/JDK-8251892

k3KAW8Pnf7mkmdSMPHz27 commented:
Not sure if you still need any input but it looks the same as for Gena928 (i.e., great)

![image](https://user-images.githubusercontent.com/2598631/90576340-ee200100-e18b-11ea-850a-f6df27beb05e.png)

Siedlerchr commented:
The manual resigning is based on the script here https://github.com/rokstrnisa/unattach/blob/master/package.sh#L47-L65

koppor commented:
I fixed the indent. Added version number to the filename (by using the available variables).

In case that works, the only thing left is https://github.com/JabRef/jabref/pull/6748#discussion_r471114125, isn't it?

koppor reviewed (changes requested):
Please fix https://github.com/JabRef/jabref/pull/6748#discussion_r471114125.

Siedlerchr commented:
I think @Gena928 did the initial design of the install dialog. Need to find the original PR.
Personally, I don't see a big problem with using our modified version.

tobiasdiez commented:
@Siedlerchr please also add this PR in #5226 as I guess most of these changes can be reverted as soon as the corresponding upstream java bugs are fixed.
